### ISSUE TITLE
[UNKNOWN_THEME_VALUE] error when referencing variables in grouping using template literal syntax

### ISSUE THREAD

**Issue by gojutin** [tw-in-js/twind#47]
This issue was originally identified in this discussion: https://github.com/tw-in-js/twind/discussions/44#discussioncomment-253093

I've included a minimal example which appears to confirm @sastan's suspicion that this issue is specific to template literal syntax. It appears to only happen when referencing outside variables.

Codesandbox Example: https://codesandbox.io/s/twind-template-literal-issue-t3zwf

**Comment by sastan** [comment]
Currently we assume that interpolation value will not affect groupings. In that case it starts a new grouping which leads to the error.

**Comment by sastan** [comment]
Here are my findings:

For performance reasons we assume that template literal strings do not change. Do that we can pre-calculate all groupings and static rules which are later combined with the dynamic rules from interpolations.

For this to work we assume that interpolations are only used as last part of a grouping. Meaning interpolations can not be used as a part of a new grouping.

**Valid**:

```js
tw`text(${'center'} underline)`
tw`text-${'center'}`
tw`hover:(text(${'center'}))`
tw`bg-${'red-600'}`
tw`bg-${'red-600'} hover:bg-${'red-700'} focus:bg-${'red-700'}`
```

**Invalid**:

```js
tw`bg-${'red'}-600`
tw`bg(${'red'}(600 700(hover:&))`
```

I currently see two ways the handle such a interpolation:

1. Fail with hint  – implemented in #56 
2. Fallback to slow parsing

@lukejacksonn What do you think?

**Comment by lukejacksonn** [comment]
This is a tough one.. part of me thinks that this should be possible:

```js
tw`bg-${'red'}-600`
```

When you say fallback to slow parsing, would it be just for these cases or for all cases? Could we detect that you are interpolating part way through a directive/grouping and then not cache that bit? This seems like a best of both worlds outcome for DX but might add some unwanted complexity?

**Comment by sastan** [comment]
With this PR we have a way to detect such grouping and throw an error. We could instead stop creating the optimization and fallback to the slow interleave mode we used to have in beamwind & oceanwind. That would mean fast for most cases and slow for this one. Maybe we could warn when we are using slow parsing.

**Comment by gojutin** [comment]
> This is a tough one.. part of me thinks that this should be possible:
> 
> ```js
> tw`bg-${'red'}-600`
> ```

I also like being able to get granular with dynamic class generation, even if it means a small performance hit. Also, this feature is another big improvement from Tailwind since certain dynamic classes in Tailwind would often lead to [unintended purging ](https://www.bing.com/search?FORM=U597DF&PC=U597&q=tailwindcss+purge)during the build step.

**Comment by sastan** [comment]
There seems to be consent that we should support this.

Should we warn when deopt?

**Comment by gojutin** [comment]
I think a warning in development env would be nice. Maybe even eventually a linting warning?

**Comment by lukejacksonn** [comment]
Yeh, like @gojutin says, I think it is one of the _big_ advantages we are afforded when taking a runtime approach. If we can get away with is without too much code disruption then we should do it.

Regards warning, I'm not too sure if it is even necessary to be honest.. when we say "slow" we still mean tens of thousands of operations per second 😅 and like you say @sastan this is not a _usual_ case.

**Comment by sastan** [comment]
I'll start to implement this as agreed on.

**Comment by sastan** [comment]
Implemented in #56 

## Try the Preview Package

> Official releases are **only** available on [registry.npmjs.org](https://www.npmjs.com/package/twind) as `twind`.

This PR has been published to [npm.pkg.github.com](https://github.com/orgs/tw-in-js/packages?repo_name=twind) as `@tw-in-js/twind@pr56`.

**Install/Update**

<details><summary>Configure your NPM client (click to expand)</summary>

Adjust you `.npmrc` to use `https://npm.pkg.github.com` for `@tw-in-js`

```
@tw-in-js:registry=https://npm.pkg.github.com
```

Using the command line:

```sh
npm config set @tw-in-js:registry https://npm.pkg.github.com --global
```

</details>

```sh
# For npm
npm install --force twind@npm:@tw-in-js/twind@pr56

# For yarn - upgrade implies install
yarn upgrade twind@npm:@tw-in-js/twind@pr56
```
